shared-module/vectorio/VectorShape.c

```diff
@@ -214,18 +214,30 @@ bool vectorio_vector_shape_fill_area(vectorio_vector_shape_t *self, const _displ
             #endif
             VECTORIO_SHAPE_PIXEL_DEBUG(" -> %d", input_pixel.pixel);
 
-            output_pixel.opaque = true;
-            if (self->pixel_shader == mp_const_none) {
-                output_pixel.pixel = input_pixel.pixel;
-            } else if (mp_obj_is_type(self->pixel_shader, &displayio_palette_type)) {
-                output_pixel.opaque = displayio_palette_get_color(self->pixel_shader, colorspace, input_pixel.pixel, &output_pixel.pixel);
-            } else if (mp_obj_is_type(self->pixel_shader, &displayio_colorconverter_type)) {
-                displayio_colorconverter_convert(self->pixel_shader, colorspace, &input_pixel, &output_pixel);
-            }
-            if (!output_pixel.opaque) {
+            // vectorio shapes use 0 to mean "area is not covered."
+            // We can skip all the rest of the work for this pixel if it's not currently covered by the shape.
+            if (input_pixel.pixel == 0) {
                 VECTORIO_SHAPE_PIXEL_DEBUG(" (encountered transparent pixel; input area is not fully covered)\n");
                 full_coverage = false;
             } else {
+                // Pixel is not transparent. Let's pull the pixel value index down to 0-base for more error-resistant palettes.
+                input_pixel.pixel -= 1;
+                output_pixel.opaque = true;
+
+                if (self->pixel_shader == mp_const_none) {
+                    output_pixel.pixel = input_pixel.pixel;
+                } else if (mp_obj_is_type(self->pixel_shader, &displayio_palette_type)) {
+                    output_pixel.opaque = displayio_palette_get_color(self->pixel_shader, colorspace, input_pixel.pixel, &output_pixel.pixel);
+                } else if (mp_obj_is_type(self->pixel_shader, &displayio_colorconverter_type)) {
+                    displayio_colorconverter_convert(self->pixel_shader, colorspace, &input_pixel, &output_pixel);
+                }
+
+                // We double-check this to fast-path the case when a pixel is not covered by the shape & not call the color converter unnecessarily.
+                if (output_pixel.opaque) {
```

Comment:
Is this check correct, or does it need to be negated with `!`?

I'm not sure if this is an enhancement or a simplification, but:

Haven't you made input_pixel be the same for every pixel now? If so, can the conversion to an output_pixel be done above, and the "double-check" becomes "if the output pixel is not opaque, *NOTHING* is drawn, and the loops can be skipped"? This also moves code out of the inner loop, so that maybe it becomes faster.

Reply:
I'm not sure why my local feather test harness is working right with that error. I swear I am a professional software developer 🤦

re input_pixel:
Look closer at what displayio_colorconverter_convert can do. This has to be in here (I tested that colorconverter transparency still works).
The perf increase here comes from skipping a function call per pixel that we know is not covered by the shape. We were computing a color that we knew we were never ~going~ supposed to consume.

Reply:
I woke up this morning and realized you meant something else about the pixel value than what I initially understood. So yeah I agree with you and will update it - I have a larger change baking atm and will wrap it into that.